### ISSUE TITLE
Same query returns different results when executed several times

### ISSUE THREAD

**Issue by ricardopereira33** [trinodb/trino#19670]
We are facing a weird issue with Trino version 428: **the same query returns different results when we run it several times.**

The process is basically 2 normal tables that were updated by Trino, and when we union them together, most of the time it returns 5 records (1 record from src table and 4 from trg table, which is correct), but sometimes it only returns 2 records (1 record from each table). 

We managed to reproduce this scenario with the SQL script (attached in the thread). If we look at the query plan (also in the thread), we can also see 2 different results, so Trino is actually processing different things, but the table is not changed. Our Trino cluster is running on GCP and reading and writing to Apache Iceberg tables stored in GCS. 

May you know what can be the issue?

<details>
  <summary> SQL script </summary>

```sql
create schema sandbox
​
​
CREATE TABLE sandbox.users_src
(
user_id uuid
,source varchar
,load_date timestamp(6)
,origin_date timestamp(6)
,first_name varchar
,second_name varchar
,third_name varchar
) 
​
CREATE TABLE sandbox.users_trg
(
user_id uuid
,source varchar
,load_date timestamp(6)
,origin_date timestamp(6)
,first_name varchar
,second_name varchar
,third_name varchar
) 
​
CREATE TABLE sandbox.users_attributes_src
(
user_id uuid
,source varchar
,load_date timestamp(6)
,origin_date timestamp(6)
,first_name varchar
,second_name varchar
,third_name varchar
) 
​
-- empty tables
select 'src' as col1, *
from sandbox.users_src
union all
select 'trg' as col1, *
from sandbox.users_trg 
​
​
-- add 1 record
insert into sandbox.users_src 
select cast ('f73894f0-5447-41c5-a727-436d04c7f8ab' as uuid), 'core.identity', current_timestamp, current_timestamp, 'John', 'Tailor', 'Machiavelli' 
​
-- add 2 record
insert into sandbox.users_src 
select cast ('4f676658-67c9-4e80-83be-ec75f0b9d0c9' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Alice', 'Tailor', '' 
​
-- add 3 record
insert into sandbox.users_src
select cast ('07001ace-3fb9-48db-8b88-26d637710b14' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Alex', 'Alex', '' 
​
-- add 4 record
insert into sandbox.users_src 
select cast ('0700390c-dfac-4896-af87-da4fb9e99a46' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Mike', 'Mike', '' 
​
​
​
insert into sandbox.users_attributes_src
select cast ('4f676658-67c9-4e80-83be-ec75f0b9d0c9' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Alice', 'Tailor', '' 	
​
insert into sandbox.users_attributes_src
select cast ('0700390c-dfac-4896-af87-da4fb9e99a46' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Mike', 'Tailor IV', ''	 
​
insert into sandbox.users_attributes_src
select cast ('0700390c-dfac-4896-af87-da4fb9e99a46' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Mike', 'Tailor II', ''	 
​
insert into sandbox.users_attributes_src
select cast ('0700390c-dfac-4896-af87-da4fb9e99a46' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Mike', 'Tailor IV', ''	 
​
insert into sandbox.users_attributes_src
select cast ('07001ace-3fb9-48db-8b88-26d637710b14' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Alex', 'Alex', '' 	
​
insert into sandbox.users_attributes_src
select cast ('f73894f0-5447-41c5-a727-436d04c7f8ab' as uuid), 'core.identity', current_timestamp, current_timestamp, 'John', 'Tailor II', 'Machiavelli' 	
​
insert into sandbox.users_attributes_src
select cast ('0700390c-dfac-4896-af87-da4fb9e99a46' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Mike', 'Mike', ''	 
​
​
-- check: 4 records in 'src'
select 'src' as col1, *
from sandbox.users_src
union all
select 'trg' as col1, *
from sandbox.users_trg 
​
-- add new records to 'trg' : + 4 records
insert into sandbox.users_trg
select *
from sandbox.users_src src
where not exists (select 1 from sandbox.users_trg trg where trg.user_id = src.user_id) 
​
-- check: 4 + 4
select 'src' as col1, *
from sandbox.users_src
union all
select 'trg' as col1, *
from sandbox.users_trg 
​
-- + 1 more record with Mike to 'src'
insert into sandbox.users_src 
select cast ('0700390c-dfac-4896-af87-da4fb9e99a46' as uuid), 'core.identity', current_timestamp, current_timestamp, 'Mike', 'Mike II', '' 
​
-- check: 5 + 4
select 'src' as col1, *
from sandbox.users_src
union all
select 'trg' as col1, *
from sandbox.users_trg 
​
-- run the same update 4-6 times on 'trg'
update sandbox.users_trg
set second_name = 'Mike II'
where user_id = cast ('0700390c-dfac-4896-af87-da4fb9e99a46' as uuid) 
​
​
-- run the "union all" query: 5 - 15 times. It should always return 1 + 4 = 5
-- you can also run each sub-query to check the output 
-- but it occasionally returns 2
select *
from 
(    
    -- 1 record
select 
      'src' as col1
    , src.user_id
    , src.source
    , src.load_date
    , src.origin_date
    , src.first_name
    , src.second_name
    , src.third_name
from sandbox.users_src as src
where not exists 
     (select 1 from sandbox.users_trg trg 
     where trg.user_id = src.user_id
     and trg.load_date = src.load_date)
union all 
    -- 4 records
select 
    'trg' as col1
    ,trg.user_id
    , trg.source
    , trg.load_date
    , trg.origin_date
    , trg.first_name
    , trg.second_name
    , trg.third_name 
from sandbox.users_trg trg
where 1=1
and exists 
    (select 1 from sandbox.users_attributes_src src
     where trg.user_id = src.user_id)
) 
```

</details>

[temp_2_rows.txt](https://github.com/trinodb/trino/files/13295352/temp_2_rows.txt)
[temp_5_rows.txt](https://github.com/trinodb/trino/files/13295353/temp_5_rows.txt)

**Comment by findepi** [comment]
Bisecting indicated this commit as the cause
```
bf15684146dd643a0d46d18c2a0841885cd36316 Change default of iceberg.file-format to PARQUET
```

i.e. ORC is apparently not affected while Parquet apparently is.

Will redo forcing format=PARQUET.

**Comment by findepi** [comment]
**TL;DR**  Looks like a bug for Parquet/UUID/DF combination of factors

cc @raunaqmorarka 

Simplest repro I could get:

```java
public class TestFooBar
        extends AbstractTestQueryFramework
{
    @Override
    protected QueryRunner createQueryRunner()
            throws Exception
    {
        return IcebergQueryRunner.builder()
                .setIcebergProperties(Map.of("iceberg.file-format", "ORC"))
                .build();
    }

    @BeforeAll
    public void setup()
    {
        assertUpdate("CREATE TABLE users_src (user_id uuid)");
        assertUpdate("CREATE TABLE users_trg (user_id uuid) WITH (format='PARQUET')");
        assertUpdate("CREATE TABLE users_attributes_src (user_id uuid)");

        assertUpdate("INSERT INTO users_src SELECT UUID 'f73894f0-5447-41c5-a727-436d04c7f8ab'", 1);
        assertUpdate("INSERT INTO users_src SELECT UUID '4f676658-67c9-4e80-83be-ec75f0b9d0c9'", 1);

        assertUpdate("INSERT INTO users_attributes_src SELECT UUID 'f73894f0-5447-41c5-a727-436d04c7f8ab'", 1);

        // Copy users_src to users_trg
        assertUpdate("INSERT INTO users_trg SELECT * FROM users_src src", 2);

        // Add one more users_src row
        assertUpdate("INSERT INTO users_src SELECT UUID 'ffffffff-dfac-4896-af87-da4fb9e99a46'", 1);
    }

    @RepeatedTest(30)
    public void test()
    {
        String query = """
                -- 1 record
                (
                    SELECT 'src' AS col1, src.user_id
                    FROM users_src src
                    WHERE NOT EXISTS (SELECT 1 FROM users_trg trg WHERE trg.user_id = src.user_id)
                )
                UNION ALL
                -- 4 records
                (
                    SELECT 'trg' AS col1, trg.user_id
                    FROM users_trg trg
                    WHERE EXISTS (SELECT 1 FROM users_attributes_src src WHERE trg.user_id = src.user_id)
                )
                """;

        Session simplifiedSession = Session.builder(getSession())
                .setSystemProperty(SystemSessionProperties.JOIN_REORDERING_STRATEGY, "NONE")
                .setSystemProperty(SystemSessionProperties.JOIN_DISTRIBUTION_TYPE, "PARTITIONED")
                .setCatalogSessionProperty("iceberg", "statistics_enabled", "false")
                //.setSystemProperty(SystemSessionProperties.ENABLE_DYNAMIC_FILTERING, "false") // no longer reproducible with DF disabled
                .build();

        MaterializedResult materializedRows = computeActual(simplifiedSession, query);

//        QueryRunner runner = getQueryRunner();
//        MaterializedResult materializedRows = TransactionBuilder.transaction(runner.getTransactionManager(), runner.getMetadata(), runner.getAccessControl())
//                .execute(simplifiedSession, session -> {
//                            Metadata metadata = runner.getMetadata();
//                            FunctionManager functionManager = runner.getFunctionManager();
//                            QueryRunner.MaterializedResultWithPlan result = runner.executeWithPlan(session, query, WarningCollector.NOOP);
//
//                            Plan plan = result.getQueryPlan();
//                            String formattedPlan = PlanPrinter.textLogicalPlan(plan.getRoot(), plan.getTypes(), metadata, functionManager, StatsAndCosts.empty(), session, 0, false);
//                            System.out.println(formattedPlan);
//
//                            return result.getMaterializedResult();
//                        });

        assertThat(materializedRows)
                .hasSize(2);
    }
}
```

**Comment by findepi** [comment]
Apparently versions as old as 380 are affected.

**Comment by raunaqmorarka** [comment]
@findepi adding `"iceberg.dynamic-filtering.wait-timeout", "10s"` to the above test will make the failure deterministic

**Comment by findepi** [comment]
Looks like same problem as https://github.com/trinodb/trino/pull/12911 but for DF pushdown.
